File: packages/runtime-dotnet/src/Validate.cs
namespace Besdk.Runtime;

/// <summary>
/// Walks a response against its declared shape (SPEC.md §3.4.1.1).
/// </summary>
/// <remarks>
/// Two things this deliberately never checks:
/// <list type="bullet">
/// <item><description>
/// <b>Unknown fields.</b> A server adding a field must not break a client. That is the whole point of an
/// evolving API.
/// </description></item>
/// <item><description>
/// <b>Enum membership.</b> Servers add enum values without warning, and the open-enum rule (§3.3.1) exists
/// precisely so that does not break a client — checking membership here would reintroduce it.
/// </description></item>
/// </list>
/// </remarks>
public static class Validate
{
  /// <summary>Collect the ways a value fails to match a schema.</summary>
  public static IReadOnlyList<string> Check(
      object? value,
      Schema schema,
      IReadOnlyDictionary<string, Schema> table)
  {
    var problems = new List<string>();
    Walk(value, schema, table, string.Empty, problems);
    return problems;
  }

  /// <summary>Throw when validation fails, honouring the mode.</summary>
  public static void Enforce(
      object? value,
      Schema schema,
      IReadOnlyDictionary<string, Schema> table,
      string operation,
      ValidationMode mode)
  {
    if (mode == ValidationMode.Off)
    {
      return;
    }

    var problems = Check(value, schema, table);
    if (problems.Count == 0)
    {
      return;
    }

    if (mode == ValidationMode.Warn)
    {
      // `Trace` rather than `Console`: a library writing to stdout corrupts anything parsing a host's
      // output, and .NET's diagnostic plumbing is what a host is already listening to.
      System.Diagnostics.Trace.TraceWarning(
          "{0}: response did not match the declared shape — {1}", operation, problems[0]);
      return;
    }

    throw new ResponseValidationException(operation, problems);
  }

  private static void Walk(
      object? value,
      Schema schema,
      IReadOnlyDictionary<string, Schema> table,
      string path,
      List<string> problems)
  {
    var where = path.Length == 0 ? "the response" : path;
    switch (schema)
    {
      case Schema.Any:
        break;
      case Schema.Str:
      case Schema.Date:
        Expect(value is string, where, "a string", value, problems);
        break;
      case Schema.Num:
        Expect(value is long or double or int or decimal, where, "a number", value, problems);
        break;
      case Schema.Int:
        // A JSON integer may arrive as a whole double from a serialiser with no integer type; rejecting
        // that would fail on data that is correct.
        Expect(IsInteger(value), where, "an integer", value, problems);
        break;
      case Schema.Bool:
        Expect(value is bool, where, "a boolean", value, problems);
        break;
      case Schema.Nullable nullable:
        if (value is not null)
        {
          Walk(value, nullable.Inner, table, path, problems);
        }

        break;
      case Schema.Arr array:
        WalkArray(value, array, table, path, where, problems);
        break;
      case Schema.MapOf mapOf:
        WalkMap(value, mapOf, table, path, where, problems);
        break;
      case Schema.Obj obj:
        WalkObject(value, obj, table, path, where, problems);
        break;
      case Schema.Or union:
        foreach (var branch in union.Branches)
        {
          if (Check(value, branch, table).Count == 0)
          {
            return;
          }
        }

        // One message rather than every branch's failure: a union of five reporting five problems buries
        // the actual one.
        Expect(false, where, "one of the declared shapes", value, problems);
        break;
      case Schema.Ref reference:
        // A missing entry is treated as `any`: an incomplete table must not reject correct data. A cycle
        // terminates here, through the table, rather than through recursion.
        if (table.TryGetValue(reference.Name, out var target))
        {
          Walk(value, target, table, path, problems);
        }

        break;
      default:
        break;
    }
  }

  private static void WalkArray(
      object? value,
      Schema.Arr array,
      IReadOnlyDictionary<string, Schema> table,
      string path,
      string where,
      List<string> problems)
  {
    if (value is not IList<object?> items)
    {
      Expect(false, where, "an array", value, problems);
      return;
    }

    for (var index = 0; index < items.Count; index++)
    {
      Walk(items[index], array.Items, table, $"{path}[{index}]", problems);
    }
  }

  private static void WalkMap(
      object? value,
      Schema.MapOf mapOf,
      IReadOnlyDictionary<string, Schema> table,
      string path,
      string where,
      List<string> problems)
  {
    // An empty map arrives as `[]` from a PHP backend, which is the artifact §3.1.2 names. A valid empty
    // map, not a wrong type.
    if (value is IList<object?> { Count: 0 })
    {
      return;
    }

    if (value is not IDictionary<string, object?> map)
    {
      Expect(false, where, "an object", value, problems);
      return;
    }

    foreach (var (key, item) in map)
    {
      Walk(item, mapOf.Values, table, Join(path, key), problems);
    }
  }

  private static void WalkObject(
      object? value,
      Schema.Obj obj,
      IReadOnlyDictionary<string, Schema> table,
      string path,
      string where,
      List<string> problems)
  {
    if (value is IList<object?> { Count: 0 })
    {
      return;
    }

    if (value is not IDictionary<string, object?> map)
    {
      Expect(false, where, "an object", value, problems);
      return;
    }

    foreach (var field in obj.Fields)
    {
      if (!map.TryGetValue(field.WireName, out var item))
      {
        if (field.Required)
        {
          problems.Add(Join(path, field.WireName) + " is missing");
        }

        continue;
      }

      Walk(item, field.Schema, table, Join(path, field.WireName), problems);
    }

    if (obj.Additional is not null)
    {
      var known = obj.Fields.Select(field => field.WireName).ToHashSet(StringComparer.Ordinal);
      foreach (var (key, item) in map)
      {
        if (!known.Contains(key))
        {
          Walk(item, obj.Additional, table, Join(path, key), problems);
        }
      }
    }
  }

  private static bool IsInteger(object? value) => value switch
  {
    long or int or short => true,
    double number => Math.Abs(number % 1) < double.Epsilon && !double.IsInfinity(number),
    _ => false,
  };

  private static void Expect(
      bool ok,
      string where,
      string expected,
      object? actual,
      List<string> problems)
  {
    if (!ok)
    {
      problems.Add($"{where} should be {expected} but was {Describe(actual)}");
    }
  }

  /// <summary>What arrived, named the way JSON names it rather than the way .NET does.</summary>
  private static string Describe(object? value) => value switch
  {
    null => "null",
    bool => "a boolean",
    long or int or short => "an integer",
    double or float or decimal => "a number",
    string => "a string",
    IList<object?> => "an array",
    IDictionary<string, object?> => "an object",
    _ => "something else",
  };

  private static string Join(string path, string segment) =>
      path.Length == 0 ? segment : path + "." + segment;
}

File: packages/runtime-java/src/main/java/besdk/runtime/Validate.java
package besdk.runtime;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Walks a response against its declared shape (SPEC.md §3.4.1.1).
 *
 * <p>Two things this deliberately never checks:
 *
 * <ul>
 *   <li><b>Unknown fields.</b> A server adding a field must not break a client. That is the whole
 *       point of an evolving API.
 *   <li><b>Enum membership.</b> Servers add enum values without warning, and the open-enum rule
 *       (§3.3.1) exists precisely so that does not break a client — checking membership here would
 *       reintroduce it.
 * </ul>
 */
public final class Validate {

  private Validate() {}

  /** Collect the ways {@code value} fails to match {@code schema}. */
  public static List<String> check(Object value, Schema schema, Map<String, Schema> table) {
    List<String> problems = new ArrayList<>();
    walk(value, schema, table, "", problems);
    return problems;
  }

  /** Throw when validation fails, honouring the mode. */
  public static void enforce(
      Object value,
      Schema schema,
      Map<String, Schema> table,
      String operation,
      ValidationMode mode) {
    if (mode == ValidationMode.OFF) {
      return;
    }
    List<String> problems = check(value, schema, table);
    if (problems.isEmpty()) {
      return;
    }
    if (mode == ValidationMode.WARN) {
      System.getLogger(Validate.class.getName())
          .log(
              System.Logger.Level.WARNING,
              operation + ": response did not match the declared shape — " + problems.get(0));
      return;
    }
    throw new ResponseValidationException(operation, problems);
  }

  private static void walk(
      Object value, Schema schema, Map<String, Schema> table, String path, List<String> problems) {
    String where = path.isEmpty() ? "the response" : path;
    // Exhaustive over a sealed type, so adding a descriptor kind without handling it is a compile
    // error.
    switch (schema) {
      case Schema.Any ignored -> {}
      case Schema.Str ignored ->
          expect(value instanceof String, where, "a string", value, problems);
      case Schema.Date ignored ->
          expect(value instanceof String, where, "a string", value, problems);
      case Schema.Num ignored ->
          expect(value instanceof Number, where, "a number", value, problems);
      case Schema.Int ignored ->
          // A JSON integer may arrive as a whole double from a serialiser with no integer type;
          // rejecting
          // that would fail on data that is correct.
          expect(isInteger(value), where, "an integer", value, problems);
      case Schema.Bool ignored ->
          expect(value instanceof Boolean, where, "a boolean", value, problems);
      case Schema.Nullable nullable -> {
        if (value != null) {
          walk(value, nullable.inner(), table, path, problems);
        }
      }
      case Schema.Arr array -> {
        if (!(value instanceof List<?> items)) {
          expect(false, where, "an array", value, problems);
          return;
        }
        for (int i = 0; i < items.size(); i++) {
          walk(items.get(i), array.items(), table, path + "[" + i + "]", problems);
        }
      }
      case Schema.MapOf mapOf -> {
        if (value instanceof List<?> empty && empty.isEmpty()) {
          // An empty map arrives as `[]` from a PHP backend, which is the artifact §3.1.2 names. A
          // valid
          // empty map, not a wrong type.
          return;
        }
        if (!(value instanceof Map<?, ?> map)) {
          expect(false, where, "an object", value, problems);
          return;
        }
        map.forEach(
            (key, item) ->
                walk(item, mapOf.values(), table, join(path, String.valueOf(key)), problems));
      }
      case Schema.Obj object -> walkObject(value, object, table, path, where, problems);
      case Schema.Or union -> {
        for (Schema branch : union.branches()) {
          if (check(value, branch, table).isEmpty()) {
            return;
          }
        }
        // One message rather than every branch's failure: a union of five reporting five problems
        // buries
        // the actual one.
        expect(false, where, "one of the declared shapes", value, problems);
      }
      case Schema.Ref ref -> {
        Schema target = table.get(ref.name());
        // A missing entry is treated as `any`: an incomplete table must not reject correct data. A
        // cycle
        // terminates here, through the table, rather than through recursion.
        if (target != null) {
          walk(value, target, table, path, problems);
        }
      }
    }
  }

  private static void walkObject(
      Object value,
      Schema.Obj object,
      Map<String, Schema> table,
      String path,
      String where,
      List<String> problems) {
    if (value instanceof List<?> empty && empty.isEmpty()) {
      // Same PHP artifact as above: `{}` serialised as `[]`.
      return;
    }
    if (!(value instanceof Map<?, ?> map)) {
      expect(false, where, "an object", value, problems);
      return;
    }
    for (Schema.Field field : object.fields()) {
      if (!map.containsKey(field.wireName())) {
        if (field.required()) {
          problems.add(join(path, field.wireName()) + " is missing");
        }
        continue;
      }
      walk(
          map.get(field.wireName()), field.schema(), table, join(path, field.wireName()), problems);
    }
    if (object.additional() != null) {
      List<String> known = object.fields().stream().map(Schema.Field::wireName).toList();
      map.forEach(
          (key, item) -> {
            if (!known.contains(String.valueOf(key))) {
              walk(item, object.additional(), table, join(path, String.valueOf(key)), problems);
            }
          });
    }
  }

  private static boolean isInteger(Object value) {
    if (value instanceof Long || value instanceof Integer || value instanceof Short) {
      return true;
    }
    return value instanceof Double d && d == Math.rint(d) && !d.isInfinite();
  }

  private static void expect(
      boolean ok, String where, String expected, Object actual, List<String> problems) {
    if (!ok) {
      problems.add(where + " should be " + expected + " but was " + describe(actual));
    }
  }

  /** What arrived, named the way JSON names it rather than the way Java does. */
  private static String describe(Object value) {
    if (value == null) {
      return "null";
    }
    if (value instanceof Boolean) {
      return "a boolean";
    }
    if (value instanceof Long || value instanceof Integer) {
      return "an integer";
    }
    if (value instanceof Number) {
      return "a number";
    }
    if (value instanceof String) {
      return "a string";
    }
    if (value instanceof List<?>) {
      return "an array";
    }
    if (value instanceof Map<?, ?>) {
      return "an object";
    }
    return "something else";
  }

  private static String join(String path, String segment) {
    return path.isEmpty() ? segment : path + "." + segment;
  }
}
